man/scr_index.1
.TH scr_index 1  "" "SCR" "SCR"
.SH NAME
scr_index \- manages status of datasets stored on parallel file system

.SH SYNOPSIS
.B "scr_index [options]"

.SH DESCRIPTION
The \fBscr_index\fR command manages datasets stored on the parallel
file system.

.SH OPTIONS
.TP
.BI "-l, --list "
List contents of index (default behavior).
.TP
.BI "-a, --add " DIR
Add dataset directory to index.
.TP
.BI "-r, --remove " DIR
Remove dataset directory from index.  Does not delete corresponding files.
.TP
.BI "-c, --current " DIR
Specify directory as restart directory.
.TP
.BI "-p, --prefix"
Specify prefix directory (defaults to current working directory).
.TP
.BI "-h, --help"
Print usage.

.LP
When adding a directory to the index, the command checks
that the dataset directory on the parallel file system contains a
complete and valid set of files, it rebuilds missing
files if possible, and it writes a \fIsummary\fR
file, which is used by the SCR library to fetch and redistribute files
to appropriate ranks upon a restart.

One may invoke the command outside of a SLURM job allocation, which is
useful to check and rebuild a dataset set in which \fBscr_postrun\fR
may have failed to complete its internal call to \fBscr_index\fR.

When listing dataset, the internal SCR dataset id is shown,
along with a flag denoting whether the dataset is valid,
the time the dataset was flushed, and the dataset directory name.
The dataset marked as \fBcurrent\fR is denoted with a * symbol.
During a restart, SCR will attempt to fetch the most recent
checkpoint starting from the current dataset.
SCR will not attempt to fetch any dataset marked as invalid.

.SH EXAMPLES
.TP
(1) List dataset direcotries in index file:
.nf
>> scr_index --list
   DSET VALID FLUSHED             DIRECTORY
*    18 YES   2015-09-28T16:46:22 scr.dataset.18
     12 YES   2015-09-28T16:43:40 scr.dataset.12
      6 YES   2015-09-28T16:43:02 scr.dataset.6
.fi
.TP
(2) Add dataset directory to index file, rebuild missing files if necessary, and create summary file:
.nf
>> scr_index --add scr.dataset.20
.fi

.SH SEE ALSO
\fISCR User Manual\fR
.LP
\fBscr\fR(1), \fBscr_srun\fR(1), \fBscr_halt\fR(1),
\fBscr_prerun\fR(1), \fBscr_postrun\fR(1)


ChangeLog
2012-02-29 Adam Moody <moody20@llnl.gov>

	* scr_util_mpi.c : Added some functions to send/recv strings via MPI.

	* scr_reddesc.c : Step through redundancy descriptor hashes by name
	as determined by rank 0 (no longer need to number CKPTDESC in
	sequential order).

2012-02-08 Adam Moody <moody20@llnl.gov>

	* m4/x_ac_expand_install_dirs.m4 : Fix to address datarootdir variable
	added from autoconf 2.59 to 2.63.

2012-02-07 Adam Moody <moody20@llnl.gov>

	* scr_param.c : Added hash to return pointers to values returned from
	getenv, return those values directly caused seg faults on cray
	systems.

	* scr_globals.c : Removed scr_hop_distance, added, scr_group, scr_comm_node,
	and scr_comm_node_across.

2012-02-01 Adam Moody <moody20@llnl.gov>

	* scr_util.c : Added scr_strdupf.

2012-01-31 Adam Moody <moody20@llnl.gov>

	* : split scr.c into component files and moved global variables to scr_globals.c

	* : Dropped scr_comm_local/level and associated parameters and replaced
	  them with group descriptors (defines high-level groups of processes)
	  and store descriptors (defines parameters about storage locations).

2011-08-31 Adam Moody <moody20@llnl.gov>

	* scr_hash_util.h/c : Added scr_hash_set/get_double.

	* scr_transfer.c : Only bother writing to the transfer file if we
	  successfully opened it.

	* scr.c : Added code to support async flush in 1.1-8.

2011-08-03 Adam Moody <moody20@llnl.gov>

	* : Recording meta data in filemap and summary files instead of separate
	  meta data files.

	* scr_filemap.h,
	  scr_filemap.c : Added routines to set/get/unset meta for each file.

2011-07-30 Adam Moody <moody20@llnl.gov>

	* : Changed filemap, summary, and index files to refer to dataset ids
	  rather than checkpoint ids.  Rebuild now rebuilds all datasets in cache
	  instead of just the most recent.  Not yet fully tested.

	* : Moved halt, flush, and nodes files from control directory to prefix
	  directory.  Now only written by rank 0 rather than all master nodes.

	* scr_scavenge.in : Renamed scr_flush to scr_scavenge and changed verbage
	  from flush and drain to scavenge where appropriate.  Flush was overloaded
	  to refer to different concepts and drain soon will be.

	* scr_hash.h,
	  scr_hash.c : Moved function to list integer keys from filemap to hash.
	  Fixed bug which was sorting integer keys as strings in scr_hash_sort_int().

2011-06-28 Adam Moody <moody20@llnl.gov>

	* scr_env.h,
	  scr_env.c : Added file to isolate typical machine dependent code.

	* scr_dataset.h,
	  scr_dataset.c : Associates attributes with a set of files.  Extends
	  SCR to handle general output datasets rather than just checkpoints.

	* scr_err.h,
	  scr_err.c,
	  scr_err_serial.c : Write all messages to stdout (errors were going to stderr).
	  Include version number in all messages.  Added new scr_warn() function.

	* scr_io.h,
	  scr_io.c : Added functions to return current working directory and
	  to build the absolute path to a file.  Also added functions to compress
	  and decompress files.

	* scr_util.h,
	  scr_util.c : Added functions to pack and unpack integer types into a
	  buffer in network order.

	* scr_hash_util.h,
	  scr_hash_util.c : Added functions to set/get integer values.

	* scr_flush_file.c : Fixed bug that was returning with wrong code if
	  named checkpoint is invalid.

	* scr_hash.c : Modified code to use new pack / unpack functions.

	* scr_meta.h,
	  scr_meta.c : Modified code to use new get / set hash functions.  Also added
	  functions to record original directory specified by user at SCR_Route_file time.

	* scr_copy.c : Allow caller to specify parameters using options instead of arg order.

	* scr_dataset.h,
	  scr_dataset.c : Added functions to set / get / unset a dataset hash.

	* scr_conf.h,
	  scr.c : Added partial support for containers (available in flush and fetch,
	  but still need drain support).  Added code to record dataset attributes.  Added
	  code to remember original directories specified by caller, need support to
	  create them and delete them with checkpoint.

2011-06-10 Adam Moody <moody20@llnl.gov>

	* scr_err.h,
	  scr_err_serial.c,
	  scr.c : Added version string to error messages.

2011-05-23 Adam Moody <moody20@llnl.gov>

	* scr_hash.c,
	  tv_data_display.c,
	  tv_data_display.h : Updated to latest C++View files to restore printing capability
	  in TotalView.

2011-04-13 Adam Moody <moody20@llnl.gov>

	* scr.c : Added scr_summary_read_v4_to_v5 to read old scr_summary.txt files and
	  convert them into equivalent v5 hash for backwards compatibility.  This enables
	  v1.1-8 to restart from checkpoints stored in v1.1-7 or earlier.

	* scr_hash.c,
	  scr_hash.h : Added scr_hash_sort to sort keys in hash by string, also modified
	  integer sort call to accept ascending or descending flag.

	* scr_index.c : Added --remove option to drop a directory from the index.

	* man/scr_index.1.in : Described format of --list output and added an example.

2011-04-06 Adam Moody <moody20@llnl.gov>

	* man/scr_index.1.in : Added man page for scr_index, and removed references to
	  scr_check_complete.

	* doc/scr_users_manual.pdf,
	  doc/scr_developers_manual.pdf : Updated user's manual and renamed it.  Added
	  developer's manual.

2011-03-24 Adam Moody <moody20@llnl.gov>

	* scr.c : Fixed bug regarding leaking group handles.

	* scr.c,
	  scr_hash_mpi.c : Moved MPI related functions for hash from scr.c to scr_hash_mpi.c.
	  Includes support for a sparse global exchange.

	* m4/x_ac_gcs.m4,
	  configure.ac : Added support for libgcs (generalized comm split library), which
	  enables one to split processes on arbitrary data.  Since this library is only at
	  LLNL currently, also added --with-gcs configure option and HAVE_LIBGCS compile flag
	  to enable it.

	* scr_hash.h,
	  scr_hash.c : Added scr_hash_getf(), similar to scr_hash_setf().
	  Added hash keys for XOR file header.

2010-12-20 Adam Moody <moody20@llnl.gov>

	* scr_io.h,

scripts/pyfe/pyfe/cli/scr_index.py
#! /usr/bin/env python3

import os

from pyfe import scr_const
from pyfe.scr_common import runproc


class SCRIndex:
  def __init__(self, prefix, verbose=False):
    self.bindir = scr_const.X_BINDIR  # path to SCR bin directory
    self.prefix = prefix  # path to SCR_PREFIX
    self.verbose = verbose
    self.exe = os.path.join(self.bindir, "scr_index") + " --prefix " + str(prefix)

  # capture and return verbatim output from the --list command
  def list(self):
    output, rc = runproc(self.exe + " --list", getstdout=True, verbose=self.verbose)
    if rc == 0:
      return output

  # make named dataset as current
  def current(self, name):
    rc = runproc(self.exe + " --current " + str(name), verbose=self.verbose)[1]
    return (rc == 0)

  # add dataset to index file (must already exist)
  def add(self, dset):
    rc = runproc(self.exe + " --add " + str(dset), verbose=self.verbose)[1]
    return (rc == 0)

  # run build command to inspect and rebuild dataset files
  def build(self, dset):
    rc = runproc(self.exe + " --build " + str(dset), verbose=self.verbose)[1]
    return (rc == 0)
